File: src/main/java/tictactoe/BoardState.java
package tictactoe;

/**
 * Represents Board State
 */
public class BoardState {
    private String line;

    public BoardState(String line) {
        if (line.length() != 9) {
            throw new IllegalArgumentException("Required exactly 9 Characters");
        }
        for (char c : line.toCharArray()) {
            if (c != 'X' && c != 'O' && c != '=') {
                throw new IllegalArgumentException("Unexpected Character - " + c);
            }
        }
        this.line = line;
    }

    /**
     * get winner
     * @return winner   null if no winner
     */
    public Player getWinner() {
        Player winner = getHorizontalWinner();
        if (winner == null) {
            winner = getVerticalWinner();
        }
        if (winner == null) {
            winner = getDiagonalWinner();
        }
        return winner;
    }

    /**
     * get next player
     * @return next player
     */
    public Player getNextPlayer() {
        int numberOfX = 0;
        int numberOfO = 0;
        for (char c : line.toCharArray()) {
            if (c == Player.X.getCharValue()) {
                numberOfX++;
            } else if (c == Player.O.getCharValue()) {
                numberOfO++;
            }
        }
        return numberOfO > numberOfX ? Player.X : Player.O;
    }

    /**
     * get next winner
     * @return next winner  null if no next winner
     */
    public Player getNextWinner() {
        Player nextPlayer = getNextPlayer();
        for (int i = 0; i < line.length(); i++) {
            if (line.charAt(i) == '=') {
                String nextPossibleBoardState = line.substring(0, i) + nextPlayer.toString() + line.substring(i + 1);
                if (new BoardState(nextPossibleBoardState).getWinner() == nextPlayer) {
                    return nextPlayer;
                }
            }
        }
        return null;
    }

    /**
     * is current game completed
     */
    public boolean isCompletedGame() {
        if (getWinner() != null) {
            return true;
        }
        for (char c : line.toCharArray()) {
            if (c == '=') {
                return false;
            }
        }
        return true;
    }

    private Player getHorizontalWinner() {
        for (int i = 0; i < 3; i++) {
            int countX = 0;
            int countO = 0;
            for (int j = 0; j < 3; j++) {
                char c = line.charAt(i * 3 + j);
                if (c == Player.X.getCharValue()) {
                    countX++;
                } else if (c == Player.O.getCharValue()) {
                    countO++;
                }
            }
            if (countX == 3) {
                return Player.X;
            }
            if (countO == 3) {
                return Player.O;
            }
        }
        return null;
    }

    private Player getVerticalWinner() {
        for (int i = 0; i < 3; i++) {
            int countX = 0;
            int countO = 0;
            for (int j = 0; j < 3; j++) {
                char c = line.charAt(j * 3 + i);
                if (c == Player.X.getCharValue()) {
                    countX++;
                } else if (c == Player.O.getCharValue()) {
                    countO++;
                }
            }
            if (countX == 3) {
                return Player.X;
            }
            if (countO == 3) {
                return Player.O;
            }
        }
        return null;
    }

    private Player getDiagonalWinner() {
        int countX1 = 0;
        int countO1 = 0;
        int countX2 = 0;
        int countO2 = 0;
        for (int i = 0; i < 3; i++) {
            char c1 = line.charAt(i * 3 + i);
            char c2 = line.charAt((2 - i) * 3 + i);
            if (c1 == Player.X.getCharValue()) {
                countX1++;
            } else if (c1 == Player.O.getCharValue()) {
                countO1++;
            }
            if (c2 == Player.X.getCharValue()) {
                countX2++;
            } else if (c2 == Player.O.getCharValue()) {
                countO2++;
            }
        }
        if (countX1 == 3 || countX2 == 3) {
            return Player.X;
        }
        if (countO1 == 3 || countO2 == 3) {
            return Player.O;
        }
        return null;
    }
}

File: src/main/kotlin/tictactoe/BoardState.kt
package tictactoe


/**
 * Represents Board State for a Tic Tac Toe Game
 * 
 *  e.g. XXX00X00X is representing a 3x3 matrix of 
 *      XXX
 *      OOX
 *      OOX
 */
class BoardState (private val line: String) {

    init {
        require(line.length == 9) {
            "Required exactly 9 Characters"
        }
        require(line.all { it == 'X' || it == 'O' || it == '='  }) {
            "Unexpected Character - ${line.find { !(it == 'X' || it == 'O' || it == '=' ) }}"
        }
    }

    /**
     * get winner
     * @return winner   null if no winner
     */
    fun getWinner(): Player? =
        getHorizontalWinner() ?: getVerticalWinner() ?: getDiagonalWinner()

    /**
     * get next player
     * @return next player
     */
    fun getNextPlayer(): Player {
        val numberOfX = line.count{ it == Player.X.charValue}
        val numberOfO = line.count{ it == Player.O.charValue}
        return if (numberOfO > numberOfX) {
            Player.X
        } else {
            Player.O
        }
    }

    /**
     * get next winner
     * @return next winner  null if no next winner
     */
    fun getNextWinner(): Player? {
        val nextPlayer = getNextPlayer()

        for (i in line.indices){
            if (line[i] == '=') {
                val nextPossibleBoardState = line.replaceRange(i, i+1, nextPlayer.toString())
                if (BoardState(nextPossibleBoardState).getWinner() == nextPlayer) {
                    return nextPlayer
                }
            }
        }

        return null
    }

    /**
     * is current game completed
     */
    fun isCompletedGame(): Boolean {
        if (getWinner() != null ) {
            return true
        }

        return !line.any { it == '=' }
    }

    private fun getHorizontalWinner(): Player? {
        for (i in 0..2){
            val horizontal = listOf(line[i*3 + 0], line[i*3 + 1], line[i*3 + 2])
            if (horizontal.count{it == Player.X.charValue } == 3){
                return Player.X
            }
            if (horizontal.count{it == Player.O.charValue } == 3){
                return Player.O
            }
        }
        return null
    }

    private fun getVerticalWinner(): Player? {
        for (i in 0..2) {
            val vertical = listOf(line[i], line[3+i], line[6+i])
            if (vertical.count{it == Player.X.charValue } == 3){
                return Player.X
            }
            if (vertical.count{it == Player.O.charValue } == 3){
                return Player.O
            }
        }
        return null
    }

    private fun getDiagonalWinner() : Player?{
        val diagonal1 = listOf(line[0], line[4], line[8])
        if (diagonal1.count{it == Player.X.charValue } == 3){
            return Player.X
        }
        if (diagonal1.count{it == Player.O.charValue } == 3){
            return Player.O
        }

        val diagonal2 = listOf(line[2], line[4], line[6])
        if (diagonal2.count{it == Player.X.charValue } == 3){
            return Player.X
        }
        if (diagonal2.count{it == Player.O.charValue } == 3){
            return Player.O
        }
        return null
    }

}
